### PR TITLE
Brings animated (torque) maps to life

### PR DESCRIPTION
**Pull request by andy-esch**
This PR brings torque maps to cartoframes. 

### Basic usage
torque-cartoframes

### Torque Category and Numeric classification
torque-cartoframes

## Modifications / enhancements
- configures `cartoframes/assets/cartoframes.html` to use the basemap instead of relying on the named maps basemap (which seems to not be directly loadable with torque named maps)
- fixes order of cartocss generated so `Map {...}` code appears before `#layer {...}`
- restricts maps to basemap + animated layer for now
- some general code cleanups
- better handling of errors with maps api POSTs
- cleans up `cumulative` options (True/False only)
- new defaults for torque styles (size = 4 as opposed to static sizes of 10 / colors of `#2752ff`)
- constructs turbo carto style, sends it to carto, gets processed version since torque can't directly handle turbo carto styles

## Restrictions
- torque maps can only be one-layered (no static layers)
- currently cannot combine size and color attributes with torque (what to do with trails, etc.)
- the torque tiles are pulled from the named map, but no raster layers are pulled, so basemaps have to be setup elsewhere. i'm currently restricting basemaps to no-label and labels on bottom for now.

## Problems still to work through
- [x] torque category maps
- [x] numeric classification
- [ ] boolean values are not working for category classification

## ToDos

- [x] add validation that time column is only numeric or timestamp
- [ ] more ad hoc tests on different datasets

cc @makella

### CODE REVIEW

**Comment by andy-esch** [comment]
Hey @michellemho, could you take a look at this and try to break it ;)? No but seriously, could you try to find bugs or edge cases that don't work that you'd expect to work? This turned out to be more complex that I thought it would be, so I want to make sure it makes sense the way it currently works. Please leave any comments code-review style :) Thanks!

**Comment by michellemho** [comment]
I'll take a look @andy-esch and do my best to break it :)

**Comment by andy-esch** [comment]
@michellemho, how's it going?

**Comment by michellemho** [comment]
@andy-esch, is the all_month_3 dataset in the cartoframes account or somewhere available? I'm also having trouble just showing a Cartoframes map in the notebook. The iframe appears but there's nothing inside... the cartoframes version I have is '0.2.3b3'.

image

**Comment by andy-esch** [comment]
You can get it here: https://eschbacher.carto.com/api/v2/sql?q=select+*+from+all_month_3&filename=all_month_3&format=csv

Hmm. Not sure why that map's not appearing without errors. Could you look at the console and share the dataset with me? Has this dataset worked before?

**Comment by michellemho** [comment]
The console error is:

``
Failed to load resource: net::ERR_NAME_NOT_RESOLVED
cartodb.js:16 cartodb.js 3.15.12
``

The dataset is just a simple points dataset of retail store in NJ. Does not have a date column; I'm just trying to get the map to appear first! https://team.carto.com/u/michellemho-carto/dataset/retail_stores_4

**Comment by andy-esch** [comment]
What happens if you do `interactive=False`?

**Comment by michellemho** [comment]
With `interactive=False`, it still doesn't load.

It looks like it's actually a typo now... "carto.co" instead of "carto.com"!
```
cartodb.js:5 GET https://michellemho-carto.carto.co/api/v1/map/named/cartoframes_ver20170406_layers1_time0_baseid1_labels0_zoom0/jsonp?config=%7B%22south%22%3A39.02117%2C%22west%22%3A-75.50404%2C%22cartocss_0%22%3A%22%23layer%5B%27mapnik%3A%3Ageometry_type%27%3D1%5D%20%7B%20%20marker-fill-opacity%3A%201%3B%20marker-fill%3A%20%235D69B1%3B%20marker-line-opacity%3A%201%3B%20marker-width%3A%2010%3B%20marker-line-width%3A%200.5%3B%20marker-line-color%3A%20%23000%3B%20marker-allow-overlap%3A%20true%3B%7D%20%23layer%5B%27mapnik%3A%3Ageometry_type%27%3D3%5D%20%7B%20%20line-comp-op%3A%20hard-light%3B%20line-width%3A%200.5%3B%20line-opacity%3A%200.25%3B%20polygon-opacity%3A%200.9%3B%20line-color%3A%20%23FFF%3B%20polygon-fill%3A%20%235D69B1%3B%20polygon-gamma%3A%200.5%3B%7D%20%23layer%5B%27mapnik%3A%3Ageometry_type%27%3D2%5D%20%7B%20%20line-width%3A%201.5%3B%20line-color%3A%20%235D69B1%3B%7D%20%22%2C%22north%22%3A41.1160526%2C%22east%22%3A-74.01422%2C%22basemap_url%22%3A%22https%3A%2F%2Fcartodb-basemaps-%7Bs%7D.global.ssl.fastly.net%2Fdark_all%2F%7Bz%7D%2F%7Bx%7D%2F%7By%7D.png%22%2C%22sql_0%22%3A%22SELECT%20*%20FROM%20retail_stores_4%22%7D&callback=_cdbc_3405907740_1 net::ERR_NAME_NOT_RESOLVED
send @ cartodb.js:5
ajax @ cartodb.js:5
(anonymous) @ cartodb.js:19
options.compressor.options.force_compress.a.length.options.MAX_GET_SIZE.JSON.stringify.config @ cartodb.js:20
_requestGET @ cartodb.js:19
_createMap @ cartodb.js:19
(anonymous) @ cartodb.js:19
setTimeout (async)
createMap @ cartodb.js:19
getTiles @ cartodb.js:20
__update @ cartodb.js:20
onAdd @ cartodb.js:20
_layerAdd @ cartodb.js:8
addLayer @ cartodb.js:7
addLayerToMap @ cartodb.js:21
(anonymous) @ cartodb.js:23
trigger @ cartodb.js:6
c @ cartodb.js:23
(anonymous) @ cartodb.js:23
(anonymous) @ cartodb.js:23
(anonymous) @ cartodb.js:6
setTimeout (async)
x.delay @ cartodb.js:6
x.defer @ cartodb.js:6
c @ cartodb.js:23
cartodb.createLayer @ cartodb.js:23
(anonymous) @ VM675 about:srcdoc:53
```

**Comment by andy-esch** [comment]
Ok, looks like it's not related directly to this PR. Could you open a ticket about this? Looks like the credentials aren't properly parsed when generating the carto.js calls (https://github.com/CartoDB/cartoframes/blob/9449ac76d7227f69b296879ebd2de9d3609d3ddd/cartoframes/context.py#L748-L763)

**Comment by michellemho** [comment]
Ohhh... I didn't have the slash at the end of the the base_url (`https://michellemho-carto.com/`) _facepalm_

File: cartoframes/layer.py [hunk 1]
```diff
@@ -305,22 +325,53 @@ def _setup(self, layers, layer_idx):
...
+                    '        ORDER BY 2 DESC',
+                    '    ) AS _wrap',
+                    ') AS __wrap',
+                    'WHERE __wrap.{col} = orig.{col}',
+                ]]).format(col=self.color, query=self.query)
+                agg_func = '\'CDB_Math_Mode({})\''.format(
+                        'cf_value_{}'.format(self.color))
+                self.scheme = {
+                        'bins': ','.join(str(i) for i in range(1, 11)),
+                        'name': 'Bold',
```
Comment (michellemho) [inline cartoframes/layer.py]: If a user wants to style a animated map by a categorical column, it seems like the only scheme allowed is 'Bold'. If a user could runs the following code, no styling change happens.
```Python
cc.map(layers=[BaseMap(),
               Layer('all_month_3',
                     time = {'column':'time', 'method':'count', 'trails':5, 'duration':60},
                     color = {'column':'magtype','scheme':styling.pastel(10)})],
       interactive=True)
```

Comment (andy-esch) [reply cartoframes/layer.py]: Great catch! I need to update that line to
```python
'name': self.scheme.get('name', 'Bold')
```

**Pull request merged**